### PR TITLE
Add non-batch deterministic encryption functions

### PR DESCRIPTION
**Pull request by giarc3**
Deterministic functions work on Fields instead of Documents because each field should have a different derivation path

### CODE REVIEW

File: src/kms/KmsApi.ts [hunk 1]
```diff
@@ -28,11 +28,35 @@ interface UnwrapKeyResponse {
...
 
+interface DeriveKeyResponse {
+    derivedKeys: DerivedKeys;
+}
+
+export type DerivedKeys = Record<string, DerivedKey[]>;
+
+export interface DerivedKey {
+    derivedKey: Base64String;
+    tenantSecretId: string; // TODO: worth using uuid type? worth using at all?
```
Comment (skeet70) [inline src/kms/KmsApi.ts:39]: We probably don't need it at all, and once changes are in the `secretNumericId` for the assignment id can change to something more useful.

**Review by skeet70 — CHANGES_REQUESTED** [review]

File: src/TenantSecurityException.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 /**
- * 0 - 199   TspServiceException
- * 200 - 299 KmsException
- * 300 - 399 SecurityEventException
- * 900 - 999 TscException
+ ** 0 - 199   TspServiceException
+ ** 200 - 299 KmsException
+ ** 300 - 399 SecurityEventException
+ ** 900 - 999 TscException
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/TenantSecurityException.ts:5]: It probably would've been more clear if I put a space between the *s. It was so that the markdown (which shows on hover in my editor) makes it a bulleted list. As it was before, it was just a single run-on line like 
```
0 - 199 TspService Exception 200 - 299 KmsException 300 - 399 SecurityEventException 900 - 999 TscException
``` 
and was hard to follow

File: src/Util.ts [hunk 3]
```diff
@@ -94,3 +95,46 @@ export interface BatchResult<T> {
...
+/**
+ * Data that has been deterministically encrypted via the customer's KMS
+ */
+export type EncryptedField = Buffer;
+
+/**
+ * Input type for deterministic encryption
+ */
+export interface DeterministicPlaintextField {
+    plaintextField: Buffer;
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/Util.ts:113]: Good catch, thanks

File: src/kms/Crypto.ts [hunk 4]
```diff
@@ -252,3 +270,131 @@ export const decryptBatchDocuments = (
...
+    const decryptedFuture = decomposeDeterministicField(encryptedField.encryptedField).flatMap((parts) => {
+        const key = derivedKeys?.find((id) => id.tenantSecretNumericId === parts.tenantSecretId);
+        const primaryKey = derivedKeys?.find((key) => key.primary);
+        if (key === undefined) {
+            return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_DECRYPT_FAILED, "Failed deterministic decryption."));
+        } else if (reencrypt && key === primaryKey) {
+            return Future.reject(
+                new TscException(TenantSecurityErrorCode.DETERMINISTIC_REENCRYPT_SAME_SECRET, "Field is already encrypted to current tenant secret.")
+            );
+        } else {
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/kms/Crypto.ts:342]: I think I agree; this was something from Colt's original design that he asked for (with a separate error code that they could handle as they want). But we should maybe discuss this in a call to make sure we're all good with it.

Until then, I'll move the check into its own function like you suggested; you're right that I was doing it here to avoid doing the decomposition and the `.find` twice, but you're also right that they should both be cheap, so I'll keep working on my premature optimization habit.

File: src/kms/KmsApi.ts (hunk 1 above)
Comment (giarc3) [reply src/kms/KmsApi.ts:39]: Yeah, I talked to Colt about removing it from the new response structure he's working on, so I'll end up removing it here as well

File: src/kms/TenantSecurityClient.ts [hunk 5]
```diff
@@ -215,4 +218,42 @@ export class TenantSecurityClient {
...
+     */
+    deterministicDecryptField = (encryptedDoc: DeterministicEncryptedField, metadata: FieldMetadata): Promise<DeterministicPlaintextField> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, [encryptedDoc.derivationPath], encryptedDoc.secretPath, metadata)
+            .flatMap((deriveResponse) => Crypto.deterministicDecryptField(encryptedDoc, deriveResponse.derivedKeys[encryptedDoc.derivationPath]))
+            .toPromise();
+
+    /**
+     * Decrypt the provided deterministically encrypted field and re-encrypt it with the current tenant secret.
+     */
+    deterministicReencryptField = (encryptedDoc: DeterministicEncryptedField, metadata: FieldMetadata): Promise<DeterministicEncryptedField> =>
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/kms/TenantSecurityClient.ts:241]: Yeah, I think we probably should. This will end up being more like the deprecated `rekeyDocument` than the `rekeyEdek`, but I think that's okay

File: src/kms/Crypto.ts [hunk 6]
```diff
@@ -16,19 +19,34 @@ import {
...
+    HEADER_META_LENGTH_LENGTH,
+    IV_BYTE_LENGTH,
+} from "./Constants";
+import {BatchUnwrapKeyResponse, BatchWrapKeyResponse, DerivedKey} from "./KmsApi";
 import {StreamingDecryption, StreamingEncryption} from "./StreamingAes";
 import * as Util from "./Util";
+
 const pPipeline = promisify(pipeline);
 const {v3DocumentHeader, SaaSShieldHeader} = ironcorelabs.proto;
+const cryptoProvider = new miscreant.PolyfillCryptoProvider();
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/kms/Crypto.ts:37]: I'm really not sure. Hopefully @coltfred will have some insight when he gets a chance to review

File: src/kms/TenantSecurityClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (giarc3) [reply src/kms/TenantSecurityClient.ts:241]: Thinking about this more, I wonder if there will be some confusion over reusing the name/concept. `rekeyDocument` is for switching to a new tenantId, but `rekeyField` would be for switching the tenantSecretId (the tenantId _can't_ change). Do you have thoughts?

[the remark this answers is not in the document]
Comment (skeet70) [reply src/kms/TenantSecurityClient.ts:242]: It's true that could be a source of confusion. I don't know if it'd be one that would bite people or not...

[the remark this answers is not in the document]
Comment (skeet70) [reply src/kms/TenantSecurityClient.ts:242]: We decided on `rotateField`. Migration will ensure you're always on the current secret unless there's been a rotation event, so that should be the only time this is called. May want to clarify that in the documentation, but the new name might be enough with the existing doc.

**Comment by giarc3** [comment]
This has changed a lot with the change to the derive keys endpoint, so I'm going to close this PR and open a new one.
